<a href="https://colab.research.google.com/github/yaswantsaiadapa/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/yaswantsaiadapa/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I have chosen the **Content Refresh / Opportunity Scoring** lane.

### Why this lane?

The goal of this project is to identify webpages that should be prioritized for content refresh using search performance, user engagement, and content freshness signals.

I chose this lane because companies often manage thousands of webpages, making it impossible to manually review every page. Instead of relying on intuition, I want to use data analysis and machine learning to build a decision-support system that ranks webpages according to their refresh priority. This allows SEO teams to spend their limited time on pages that are most likely to improve search performance after being updated.

This lane combines exploratory data analysis, feature engineering, machine learning, explainability, and business decision-making, which closely aligns with my goal of becoming a Data Scientist and Machine Learning Engineer.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

## Decision

Help SEO managers prioritize which webpages should be refreshed first by analyzing search performance, user engagement, and content freshness signals.

## Who Acts on It?

The recommendations are used by SEO managers to prioritize work, while the SEO and content teams carry out the actual content updates.

## Action

The SEO team reviews the highest-priority webpages and refreshes them by updating outdated information, improving content quality, optimizing headings and keywords, enhancing metadata, or restructuring the page to better match current search intent.

## Cost of a Wrong Recommendation

If the model recommends refreshing low-value pages, the company wastes editor time, development effort, and business resources. If it fails to identify important declining pages, the business may lose search visibility, organic traffic, user engagement, potential customers, and revenue.

## Why Machine Learning?

Large organizations often manage thousands of webpages, making manual analysis impractical. Each page must be evaluated using multiple signals such as impressions, CTR, average position, engagement, content age, and trend. Machine learning can analyze these signals together and produce a ranked list of webpages that are most likely to benefit from content refresh, helping SEO teams make faster, more consistent, and data-driven decisions.

### Success Criteria

This project will be considered successful if it produces a prioritized list of webpages that helps SEO teams identify high-impact refresh opportunities more effectively than simple manual rules or intuition alone.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [12]:
import os
print(os.getcwd())

/content/flyrank-ml-internship-starter


In [13]:
import os
import subprocess
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"
if not os.path.isdir(REPO_DIR):
    subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
os.chdir(REPO_DIR)
print("Current Directory:", os.getcwd())

Current Directory: /content/flyrank-ml-internship-starter/flyrank-ml-internship-starter


In [14]:
import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

In [15]:
print("Dataset Overview:")
print(f"Number of webpages: {df.shape[0]}")
print(f"Number of features : {df.shape[1]}")


Dataset Overview:
Number of webpages: 30000
Number of features : 44


In [16]:
print("Average Engagement Rate by Trend Direction:")
print(df.groupby("trend_direction")["engagement_rate"].mean().round(2))

Average Engagement Rate by Trend Direction:
trend_direction
down      2.44
flat      1.37
new       2.14
stable    2.84
up        2.99
Name: engagement_rate, dtype: float64


In [17]:
corr = df["search_volume"].corr(df["impressions_90d"])
print("Correlation between Search Volume and Impressions")
print(f"{corr:.3f}")

Correlation between Search Volume and Impressions
0.001


### Observations

The starter dataset contains **30,000 webpages** described by **44 features**, providing sufficient information to analyze search performance, user engagement, and content quality at scale.

The average engagement rate varies across trend categories:

| Trend Direction | Average Engagement Rate |
|-----------------|------------------------:|
| Down | 2.44 |
| Flat | 1.37 |
| Stable | 2.84 |
| Up | 2.99 |

This suggests that pages with improving performance generally have higher user engagement than declining pages.

The correlation between **search_volume** and **impressions_90d** is approximately **0.001**, indicating that high keyword search volume alone does not guarantee high search visibility. Multiple factors influence webpage performance.

### Why this supports my lane

These observations show that webpage performance cannot be explained using a single feature. Search visibility, user engagement, content freshness, and other signals interact in complex ways. This makes **Content Refresh Opportunity Scoring** a suitable machine learning problem, where multiple signals can be analyzed together to prioritize webpages that are most likely to benefit from content updates.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

This project uses observable search performance, engagement, and content freshness signals to identify webpages that may benefit from content refresh. The results should be interpreted as **observed patterns** and **directional evidence** that can support SEO decision-making.

The recommendations produced by this project are intended as **decision-support tools**, helping SEO managers prioritize webpages for review rather than making decisions automatically.

This project **does not** predict Google's ranking algorithm, reverse engineer how Google ranks webpages, or prove that one feature directly causes another. Any relationships discovered represent patterns within the available dataset and should not be interpreted as causal proof.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.